flang/lib/Lower/IntrinsicCall.cpp

```diff
@@ -664,12 +699,96 @@ static std::string getIntrinsicWrapperName(const llvm::StringRef &intrinsic,
   return name;
 }
 
+/// fir::ExtendedValue to mlir::Value translation layer
+
+fir::ExtendedValue toExtendedValue(mlir::Value val,
+                                   Fortran::lower::FirOpBuilder &builder,
+                                   mlir::Location loc) {
+  assert(val && "optional unhandled here");
+  auto type = val.getType();
+  auto base = val;
+  mlir::Value len;
+  auto indexType = builder.getIndexType();
+  llvm::SmallVector<mlir::Value, 2> extents;
+
+  if (auto refType = type.dyn_cast<fir::ReferenceType>())
+    type = refType.getEleTy();
+
+  if (auto arrayType = type.dyn_cast<fir::SequenceType>()) {
+    type = arrayType.getEleTy();
+    llvm::ArrayRef<fir::SequenceType::Extent> shape = arrayType.getShape();
+    if (auto charType = type.dyn_cast<fir::CharacterType>()) {
+      auto cstLen = arrayType.getShape()[0];
+      shape = shape.drop_front();
```

Comment:
We probably need a cleanup pass on some of this character special handling stuff. It's diffusing out through the bridge again, it seems (observation from chasing CHAR related bugs). Collecting little idioms and moving them to the character expr helper area, maybe.

Reply:
Good point, I move character related part to the helper where I shared the code with `toDataLengthPair`.